core/node/da_dispatcher/src/da_dispatcher.rs

```diff
@@ -0,0 +1,199 @@
+use std::{future::Future, time::Duration};
+
+use anyhow::Context;
+use chrono::{NaiveDateTime, Utc};
+use rand::Rng;
+use tokio::sync::watch;
+use zksync_config::DADispatcherConfig;
+use zksync_da_client::{types::DAError, DataAvailabilityClient};
+use zksync_dal::{ConnectionPool, Core, CoreDal};
+use zksync_types::L1BatchNumber;
+
+use crate::metrics::METRICS;
+
+#[derive(Debug)]
+pub struct DataAvailabilityDispatcher {
+    client: Box<dyn DataAvailabilityClient>,
+    pool: ConnectionPool<Core>,
+    config: DADispatcherConfig,
+}
+
+impl DataAvailabilityDispatcher {
+    pub fn new(
+        pool: ConnectionPool<Core>,
+        config: DADispatcherConfig,
+        client: Box<dyn DataAvailabilityClient>,
+    ) -> Self {
+        Self {
+            pool,
+            config,
+            client,
+        }
+    }
+
+    pub async fn run(self, mut stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let pool = self.pool.clone();
+        loop {
+            if *stop_receiver.borrow() {
+                tracing::info!("Stop signal received, da_dispatcher is shutting down");
```

Comment:
Nit: You can move this log outside of the outer `loop`, so that you don't need to duplicate it for `tokio::time::timeout`.